test/integration/query/additional.js

```diff
@@ -978,6 +978,79 @@ module.exports = function (knex) {
         });
     });
 
+    it('.timeout(ms, {cancel: true}) should cancel all parallel slow queries run in a single transaction', async function () {
+      // Only mysql/postgres query cancelling supported for now
+      if (!isMysql(knex) && !isPostgreSQL(knex)) {
+        return this.skip();
+      }
+
+      const testQueries = {
+        [drivers.PostgreSQL]: function () {
+          return knex.raw('SELECT pg_sleep(10)');
+        },
+        [drivers.MySQL]: function () {
+          return knex.raw('SELECT SLEEP(10)');
+        },
+        [drivers.MySQL2]: function () {
+          return knex.raw('SELECT SLEEP(10)');
+        },
+        [drivers.MsSQL]: function () {
+          return knex.raw("WAITFOR DELAY '00:00:10'");
+        },
+        [drivers.Oracle]: function () {
+          return knex.raw('begin dbms_lock.sleep(10); end;');
+        },
+      };
+
+      const driverName = knex.client.driverName;
+      if (!Object.prototype.hasOwnProperty.call(testQueries, driverName)) {
+        throw new Error('Missing test query for dialect: ' + driverName);
+      }
+
+      const query = testQueries[driverName]();
+
+      const getProcessesForDriver = {
+        pg: async () => {
+          const results = await knex.raw('SELECT * from pg_stat_activity');
+          return _.map(_.filter(results.rows, {state: 'active'}), 'query');
+        },
+        mysql: async () => {
+          const results = await knex.raw('SHOW PROCESSLIST');
+          return _.map(results[0], 'Info');
+        },
+        mysql2: async () => {
+          const results = await knex.raw('SHOW PROCESSLIST');
+          return _.map(results[0], 'Info');
+        },
+      };
+
+      if (!Object.prototype.hasOwnProperty.call(getProcessesForDriver, driverName)) {
+        throw new Error('Missing test query for driverName: ' + driverName);
+      }
+
+      const getProcesses = getProcessesForDriver[driverName];
+
+      return knex.transaction(async (trx) => {
+        const promise = Promise.all(_.range(5).map(async () => {
+          await query.timeout(50, { cancel: true }).transacting(trx);
+        }));
+
+        await delay(10);
+        const processesBeforeTimeout = await getProcesses();
+        expect(processesBeforeTimeout).to.include(query.toString());
+
+        await expect(promise).to.eventually.be.rejected.and.to.deep.include({
```

Comment:
not sure about this part, I feel like this test will be green for current master too. Maybe execute 5 queries & ensure what from n-th query all queries (which amount >= 2) are cancelled?  

Also, please add test from this https://github.com/knex/knex/issues/4415#issuecomment-814945828

Reply:
@maximelkin You are rignt that check `await expect(promise).to.eventually.be.rejected.and.to.deep.include({...})` will pass on the current master as well.

The important check in this test is bellow:

```
        await delay(10);
        const processesAfterTimeout = await getProcesses();
        expect(processesAfterTimeout).to.not.include(query.toString());
```

Which is ensuring that a few milliseconds after the timeout *all* queries are cancelled, i.e. none of those 5 queries is running in DB any more. If you ran this test on the current master it will fail on this later check.

> Also, please add test from this #4415 (comment)

Yep.